sdks/python/apache_beam/ml/inference/base.py

```diff
@@ -243,70 +244,273 @@ def share_model_across_processes(self) -> bool:
+            'overriding this KeyedModelHandler and defining the '
+            'batch_elements_kwargs function.',
+            hints,
+            batch_kwargs)
+      env_vars = mh._env_vars
+      if len(hints) > 0:
+        logging.warning(
+            'mh %s defines the following _env_vars %s '
+            'which will be ignored. _env_vars are not respected when '
+            'more than one model handler is used in a KeyedModelHandler. '
+            'If you need env vars set at inference time, you can do so with '
+            'a custom inference function.',
+            mh,
+            env_vars)
+
+      if len(keys) == 0:
+        raise ValueError(
+            f'Empty list maps to model handler {mh}. All model handlers must '
+            'have one or more associated keys.')
+      self._id_to_mh_map[keys[0]] = mh
+      for key in keys:
+        if key in self._key_to_id_map:
+          raise ValueError(
+              f'key {key} maps to multiple model handlers. All keys must map '
+              'to exactly one model handler.')
+        self._key_to_id_map[key] = keys[0]
+
+  def load_model(self) -> Union[ModelT, _ModelManager]:
+    if not self._many_models:
+      return self._unkeyed.load_model()
+    return _ModelManager(self._id_to_mh_map)
 
   def run_inference(
       self,
       batch: Sequence[Tuple[KeyT, ExampleT]],
-      model: ModelT,
+      model: Union[ModelT, _ModelManager],
       inference_args: Optional[Dict[str, Any]] = None
   ) -> Iterable[Tuple[KeyT, PredictionT]]:
-    keys, unkeyed_batch = zip(*batch)
-    return zip(
-        keys, self._unkeyed.run_inference(unkeyed_batch, model, inference_args))
+    if not self._many_models:
+      keys, unkeyed_batch = zip(*batch)
+      return zip(
+          keys,
+          self._unkeyed.run_inference(unkeyed_batch, model, inference_args))
+
+    batch_by_key = {}
+    for key, example in batch:
+      if key not in batch_by_key:
+        batch_by_key[key] = []
+      batch_by_key[key].append(example)
+
+    predictions = []
+    for key, unkeyed_batches in batch_by_key.items():
+      id = self._key_to_id_map[key]
+      mh = self._id_to_mh_map[id]
+      keyed_model_tag = model.load(id)
+      keyed_model_shared_handle = multi_process_shared.MultiProcessShared(
+          mh.load_model, tag=keyed_model_tag)
+      keyed_model = keyed_model_shared_handle.acquire()
+      for inf in mh.run_inference(unkeyed_batches, keyed_model, inference_args):
+        predictions.append((key, inf))
+      keyed_model_shared_handle.release(keyed_model)
+
+    return predictions
 
   def get_num_bytes(self, batch: Sequence[Tuple[KeyT, ExampleT]]) -> int:
-    keys, unkeyed_batch = zip(*batch)
-    return len(pickle.dumps(keys)) + self._unkeyed.get_num_bytes(unkeyed_batch)
+    if not self._many_models:
+      keys, unkeyed_batch = zip(*batch)
+      return len(
+          pickle.dumps(keys)) + self._unkeyed.get_num_bytes(unkeyed_batch)
+    return len(pickle.dumps(batch))
```

Comment:
i wonder why this wouldn't work for all modelhandlers. but as is, we'd have an (undocumented?) difference in behavior that we don't delegate estimation to corresponding MHs.

Reply:
I think some model handlers can provide tight bounds on batch sizes without doing a full pickling of them. I added shelling out to model handlers as an improvement in #27628 since I think we can may be able to do that per key more efficiently. Or we can just use any of the model handlers since they're all taking the same input type anyways